### PR TITLE
Add experiment support

### PR DESCRIPTION
**Pull request by JoshuaKrstic**
Add experiment support.
This change:
  - pulls the experiment binary from a GCS bucket (not open-source-able)
  - gets metadata from the MDS for the experiment binary
  - calls experiment binary with metadata
  - uses new experiment data in container-runner.go

### CODE REVIEW

File: launcher/image/sync_experiments.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
```
Comment (alexmwu) [inline launcher/image/sync_experiments.sh:2]: you should `set -euo pipefail` here

File: launcher/container_runner.go [hunk 2]
```diff
@@ -49,13 +50,15 @@ type ContainerRunner struct {
 }
 
 const (
-	// hostTokenPath defined the directory in the host that will store attestation tokens
-	hostTokenPath = "/tmp/container_launcher/"
+	// tmpContainerLauncherPath defined the directory in the host that will store attestation tokens and experiment data.
```
Comment (alexmwu) [inline launcher/container_runner.go:53]: please do this rename in a separate PR, ideally putting it in a const package. @jkl73 was planning on doing this.

File: launcher/container_runner.go [hunk 3]
```diff
@@ -505,6 +508,16 @@ func (r *ContainerRunner) Run(ctx context.Context) error {
 		return fmt.Errorf("failed to fetch and write OIDC token: %v", err)
 	}
 
+	e := experiments.NewExperiments(r.logger)
```
Comment (alexmwu) [inline launcher/container_runner.go:511]: this callpoint should probably go somewhere in `main.go` and you should put the experiment flag settings in the LaunchSpec.

File: launcher/experiments/experiments.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
+// Package experiments contains functionalities to retrieve syned experiments
```
Comment (alexmwu) [inline launcher/experiments/experiments.go:1]: "Package experiments contains functionalities to retrieve synced experiments"

File: launcher/experiments/experiments.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	"io"
+	"log"
+	"os"
+)
+
+// Experiments represents a collection of experiments that were synced.
+// All failures during the reading of experiment data are logged and returned.
+// This struct intends to be as safe as possible. Errors default to experiments being off, or not
+// existing. Failures should not cascade from this struct.
+// When using this class, if an experiment is not found, it should be assumed it is off.
```
Comment (alexmwu) [inline launcher/experiments/experiments.go:16]: I don't think this line is needed, since we return an error from GetExperiment.

File: launcher/experiments/experiments.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+type Experiments struct {
+	experiments map[string]any
+	logger      *log.Logger
+}
+
+// NewExperiments returns an experiments.
+func NewExperiments(logger *log.Logger) Experiments {
```
Comment (alexmwu) [inline launcher/experiments/experiments.go:23]: Name this "New" go/go-style/decisions#repetitive-with-package

File: launcher/experiments/experiments.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ReadJSONInput  takes a reader and unmarshalls the contents into the experiments map.
+// If the unmarsahlling fails, the experiments map is set to an empty map.
+func (e *Experiments) ReadJSONInput(read io.Reader) error {
```
Comment (alexmwu) [inline launcher/experiments/experiments.go:62]: This doesn't need to be exported.

File: launcher/experiments/experiments.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		return nil, fmt.Errorf("experiment not found for key %s", key)
+	}
+
+	return val, nil
+}
+
+// ReadExperimentsFile takes a filepath, opens the file, and calls ReadJsonInput with the contents
+// of the file.
+// If the file cannot be opened, the experiments map is set to an empty map.
+func (e *Experiments) ReadExperimentsFile(fpath string) error {
```
Comment (alexmwu) [inline launcher/experiments/experiments.go:43]: Why doesn't the `New` function take the reader and do the initialization? This function seems unnecessarily exported.

File: launcher/experiments/experiments.go (hunk 4 above)
Comment (alexmwu) [inline launcher/experiments/experiments.go:1]: This can probably go in the internal directory.

File: launcher/image/preload.sh [hunk 9]
```diff
@@ -2,11 +2,19 @@
...
+readonly EXPERIMENTS_SCRIPT="experiments_sync"
+readonly GCS_BUCKET="gs://confidential-space-images_third-party_test"
 
 copy_launcher() {
   cp launcher "${CS_PATH}/cs_container_launcher"
 }
 
+copy_exeriment_client() {
+  gcloud storage cp ${GCS_BUCKET} "${CS_PATH}/${EXPERIMENTS_SCRIPT}"
+  chmod +x "${CS_PATH}/${EXPERIMENTS_SCRIPT}"
```
Comment (alexmwu) [inline launcher/image/preload.sh:14]: I think this may be why the script isn't running: it should go after the cp.

**Review by alexmwu — CHANGES_REQUESTED** [review]

File: launcher/image/sync_experiments.sh [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+readonly OEM_PATH='/usr/share/oem'
+readonly CS_PATH="${OEM_PATH}/confidential_space"
+readonly EXPERIMENTS_SCRIPT="experiments_sync"
+readonly SYNC_RESULT_FILE="/tmp/container_launcher/experiment_data"
+
+getImageDetails() {
+    # image will look like `projects/confidential-space-images/global/images/confidential-space-hardened-230600`
+    #                   or `projects/confidential-space-images/global/images/confidential-space-debug-230600`
```
Comment (alexmwu) [inline launcher/image/sync_experiments.sh:10]: Using the image name won't work for images created by our CI. It's not a big issue since we can copy an image over, but it is more work. Is there a reason we didn't go with something like build time?

File: launcher/image/preload.sh [hunk 11]
```diff
@@ -2,11 +2,19 @@
...
 readonly OEM_PATH='/usr/share/oem'
 readonly CS_PATH="${OEM_PATH}/confidential_space"
+readonly EXPERIMENTS_SCRIPT="experiments_sync"
+readonly GCS_BUCKET="gs://confidential-space-images_third-party_test"
 
 copy_launcher() {
   cp launcher "${CS_PATH}/cs_container_launcher"
 }
 
+copy_exeriment_client() {
```
[the remark this answers is not in the document]
Comment (JoshuaKrstic) [reply launcher/image/preload.sh]: Thanks for catching, fixed.

File: launcher/experiments/experiments.go (hunk 4 above)
Comment (JoshuaKrstic) [reply launcher/experiments/experiments.go:1]: Done

File: launcher/experiments/experiments.go (hunk 8 above)
Comment (JoshuaKrstic) [reply launcher/experiments/experiments.go:43]: I wanted this class to contain the logic for opening the experiment file to be able to easily suppress errors.
To be able to test this code, I needed to pass in a reader.

This was my solution to accomplish both: I can test the reader method and consumers don't have to convert the file input into a reader to use this class.

File: launcher/image/preload.sh (hunk 9 above)
Comment (JoshuaKrstic) [reply launcher/image/preload.sh]: Fixed this previously, thanks for catching this.

File: launcher/image/sync_experiments.sh (hunk 10 above)
Comment (JoshuaKrstic) [reply launcher/image/sync_experiments.sh:10]: Yes I did notice this in my testing.
We should talk about this offline because there's no perfect solution here.

File: launcher/experiments/experiments.go (hunk 8 above)
Comment (alexmwu) [reply launcher/experiments/experiments.go]: You can still test in the same package (`experiments_test.go`) without exporting. `New` can take a filepath and then call `readExperimentsFile`.

Comment (JoshuaKrstic) [reply launcher/experiments/experiments.go]: Gotcha, made the change.

File: launcher/image/sync_experiments.sh (hunk 10 above)
Comment (JoshuaKrstic) [reply launcher/image/sync_experiments.sh:12]: Will change how our presubmit images are generated as we talked about offline.
